File: backend/deal_pipeline/oms_schema.py
"""OMS staging-CSV schema: field definitions, value validation and CSV serialization.

The schema itself is ``oms_fields.json`` next to this module. The frontend carries a byte-for-byte
mirror (``chatbot-app/frontend/src/lib/pipeline/omsFields.json``) and a TypeScript port of these
helpers (``omsSchema.ts``); a test asserts the JSON copies match and the CSV/validation behaviour
here is kept identical to the TypeScript so the BFF's regenerated CSV (after an edit) and the
Lambda's CSV (after a parse) are the same bytes for the same fields.

Formats by ``type`` are listed in ``docs/deal-pipeline-design.md`` section 5.
"""

import csv
import io
import json
import re
from pathlib import Path

_SCHEMA_PATH = Path(__file__).with_name("oms_fields.json")

SCHEMA: dict = json.loads(_SCHEMA_PATH.read_text(encoding="utf-8"))
SECTIONS: list[str] = list(SCHEMA["sections"])
FIELDS: list[dict] = list(SCHEMA["fields"])
FIELD_KEYS: list[str] = [f["key"] for f in FIELDS]
FIELD_LABELS: list[str] = [f["label"] for f in FIELDS]

_BY_KEY: dict[str, dict] = {f["key"]: f for f in FIELDS}
_KEY_BY_LABEL: dict[str, str] = {f["label"]: f["key"] for f in FIELDS}

# Keep identical to FORMAT_PATTERNS in chatbot-app/frontend/src/lib/pipeline/omsSchema.ts.
FORMAT_PATTERNS: dict[str, re.Pattern] = {
    "date": re.compile(r"^(1[0-2]|[1-9])/(3[01]|[12][0-9]|[1-9])/\d{4}$"),
    "time": re.compile(r"^(1[0-2]|[1-9])(:[0-5]\d)?(AM|PM)$"),
    "percent": re.compile(r"^-?\d+\.\d{3}%$"),
    "mm": re.compile(r"^\d+\.\d{3}$"),
    "price": re.compile(r"^\d+\.\d{3}$"),
    "integer": re.compile(r"^\d+$"),
    "boolean": re.compile(r"^(Yes|No)$"),
}


def by_key(key: str) -> dict | None:
    """Return the field definition for ``key``, or None when the schema has no such field."""
    return _BY_KEY.get(key)


def key_for_label(label: str) -> str | None:
    """Return the field key whose CSV column label is ``label``, or None for an unknown label."""
    return _KEY_BY_LABEL.get(label)


def fields_by_section() -> list[tuple[str, list[dict]]]:
    """Group the fields by section, sections in schema order, fields in schema order."""
    return [(s, [f for f in FIELDS if f["section"] == s]) for s in SECTIONS]


def format_hint(field_type: str) -> str:
    """Human-readable description of the format a ``type`` expects (matches the TS ``formatHint``)."""
    return {
        "date": "M/D/YYYY",
        "time": "h[:mm]AM|PM, e.g. 12PM",
        "percent": "0.000%",
        "mm": "millions with 3 decimals, e.g. 500.000",
        "price": "3 decimals, e.g. 99.500",
        "integer": "a whole number",
        "boolean": "Yes or No",
    }.get(field_type, "text")


def validate_value(field: dict, raw) -> str | None:
    """Validate one value against its field definition.

    Messages are deliberately the same short phrases the TypeScript ``validateFieldValue`` emits,
    so a problem shown in the review UI and a problem the agent sees read the same.

    :param field: a schema field definition (see :func:`by_key`).
    :param raw: the candidate value; None and non-strings are coerced to text first.
    :returns: None when valid, otherwise a short human-readable problem.
    """
    value = ("" if raw is None else str(raw)).strip()
    if value == "":
        return "required" if field.get("required") else None
    ftype = field["type"]
    if ftype == "enum":
        values = field.get("values") or []
        return None if value in values else f"must be one of {', '.join(values)}"
    if ftype == "string":
        max_length = field.get("max_length")
        if max_length and len(value) > max_length:
            return f"longer than {max_length} characters"
        pattern = field.get("pattern")
        if pattern and not re.search(pattern, value):
            return f"must match {pattern}"
        return None
    if ftype == "integer":
        if not FORMAT_PATTERNS["integer"].match(value):
            return "must be a whole number"
        n = int(value)
        if field.get("min") is not None and n < field["min"]:
            return f"must be ≥ {field['min']}"
        if field.get("max") is not None and n > field["max"]:
            return f"must be ≤ {field['max']}"
        return None
    pattern = FORMAT_PATTERNS.get(ftype)
    if pattern and not pattern.match(value):
        return f"expected {format_hint(ftype)}"
    return None


def empty_fields() -> dict[str, str]:
    """Every key present, blanks as "", schema defaults applied."""
    return {f["key"]: f.get("default", "") for f in FIELDS}


def normalize_fields(values: dict | None) -> dict[str, str]:
    """Fill missing keys with defaults/blanks, stringify values and drop unknown keys.

    Mirrors the TS ``normalizeFields``: an explicit "" in the input overrides a schema default,
    because the caller (an editor, or the model) said the field is blank. Use
    :func:`apply_defaults` afterwards when blanks should fall back to the defaults.

    :param values: partial mapping of field key to value; None means "nothing".
    :returns: a complete field mapping in schema key order.
    """
    out = empty_fields()
    for key, value in (values or {}).items():
        if key in _BY_KEY:
            out[key] = "" if value is None else str(value)
    return out


def apply_defaults(fields: dict[str, str]) -> dict[str, str]:
    """Return a copy of ``fields`` with schema defaults filled into blank values."""
    out = dict(fields)
    for f in FIELDS:
        if "default" in f and not (out.get(f["key"]) or "").strip():
            out[f["key"]] = f["default"]
    return out


def validate_fields(fields: dict[str, str]) -> dict[str, str]:
    """All problems across a record keyed by field key; empty when the record is clean."""
    problems = {}
    for f in FIELDS:
        problem = validate_value(f, fields.get(f["key"], ""))
        if problem:
            problems[f["key"]] = problem
    return problems


_NEEDS_QUOTING = re.compile(r'[",\r\n]')


def _csv_cell(value) -> str:
    """RFC 4180 quoting exactly as the TS ``csvCell``: quote only when the cell needs it."""
    s = "" if value is None else str(value)
    if _NEEDS_QUOTING.search(s):
        return '"' + s.replace('"', '""') + '"'
    return s


def to_csv(fields: dict[str, str]) -> str:
    """Header row of labels plus one data row, "\\n" line endings, RFC 4180 quoting.

    Byte-identical to the TypeScript ``toCsv`` for the same field values, which is what lets the
    BFF regenerate a deal's CSV after an edit without the two sides drifting.
    """
    header = ",".join(_csv_cell(f["label"]) for f in FIELDS)
    row = ",".join(_csv_cell(fields.get(f["key"], "")) for f in FIELDS)
    return f"{header}\n{row}\n"


def parse_csv(text: str) -> tuple[list[str], dict[str, str]]:
    """Read a one-record staging CSV back into ``(header_labels, fields_by_key)``.

    The header is returned verbatim so the OMS validator can report exactly how it differs from
    the schema (``HEADER_MISMATCH``); values are keyed by field key for every header label the
    schema knows, so a misordered file still yields values to validate. Only structural
    problems raise here — a file the OMS could not even read as "one header row, one data row".

    :param text: the CSV text.
    :returns: the header labels in file order, and the known values keyed by field key.
    :raises ValueError: when the file has no header row, no data row, or the data row's cell
        count differs from the header's.
    """
    rows = [r for r in csv.reader(io.StringIO(text)) if r]
    if len(rows) < 2:
        raise ValueError("CSV must contain a header row and one data row")
    labels, data = rows[0], rows[1]
    if len(data) != len(labels):
        raise ValueError(f"data row has {len(data)} cells but the header has {len(labels)}")
    fields = {}
    for label, value in zip(labels, data):
        key = _KEY_BY_LABEL.get(label)
        if key is not None:
            fields[key] = value
    return labels, fields

File: chatbot-app/frontend/src/lib/pipeline/omsSchema.ts
// OMS staging-CSV schema helpers shared by the BFF (server) and the UI (client).
// The schema itself is backend/deal_pipeline/oms_fields.json, mirrored here as omsFields.json;
// a test asserts the two copies are identical. Formats: docs/deal-pipeline-design.md §5.

import schemaJson from "./omsFields.json";
import type { FieldValues, OmsFieldDef, OmsSchema } from "./types";

export const OMS_SCHEMA: OmsSchema = schemaJson as OmsSchema;
export const OMS_FIELDS: OmsFieldDef[] = OMS_SCHEMA.fields;
export const OMS_SECTIONS: string[] = OMS_SCHEMA.sections;

const BY_KEY = new Map(OMS_FIELDS.map((f) => [f.key, f]));

export function fieldDef(key: string): OmsFieldDef | undefined {
  return BY_KEY.get(key);
}

/** Fields grouped in section order, each section in schema order. */
export function fieldsBySection(): { section: string; fields: OmsFieldDef[] }[] {
  return OMS_SECTIONS.map((section) => ({
    section,
    fields: OMS_FIELDS.filter((f) => f.section === section),
  }));
}

/** Every key present, blanks as "", schema defaults applied. */
export function emptyFields(): FieldValues {
  const out: FieldValues = {};
  for (const f of OMS_FIELDS) out[f.key] = f.default ?? "";
  return out;
}

/** Fill missing keys with defaults/blanks and drop unknown keys. */
export function normalizeFields(input: Partial<FieldValues> | undefined): FieldValues {
  const out = emptyFields();
  if (!input) return out;
  for (const [k, v] of Object.entries(input)) {
    if (BY_KEY.has(k)) out[k] = v == null ? "" : String(v);
  }
  return out;
}

// Format regexes — keep identical to backend/deal_pipeline/oms_schema.py.
export const FORMAT_PATTERNS: Record<string, RegExp> = {
  date: /^(1[0-2]|[1-9])\/(3[01]|[12][0-9]|[1-9])\/\d{4}$/,
  time: /^(1[0-2]|[1-9])(:[0-5]\d)?(AM|PM)$/,
  percent: /^-?\d+\.\d{3}%$/,
  mm: /^\d+\.\d{3}$/,
  price: /^\d+\.\d{3}$/,
  integer: /^\d+$/,
  boolean: /^(Yes|No)$/,
};

/**
 * Validate one value against its field definition.
 *
 * @returns null when valid, otherwise a short human-readable problem.
 */
export function validateFieldValue(def: OmsFieldDef, raw: string): string | null {
  const value = (raw ?? "").trim();
  if (value === "") return def.required ? "required" : null;
  switch (def.type) {
    case "enum":
      return def.values?.includes(value)
        ? null
        : `must be one of ${def.values?.join(", ")}`;
    case "string": {
      if (def.max_length && value.length > def.max_length)
        return `longer than ${def.max_length} characters`;
      if (def.pattern && !new RegExp(def.pattern).test(value))
        return `must match ${def.pattern}`;
      return null;
    }
    case "integer": {
      if (!FORMAT_PATTERNS.integer.test(value)) return "must be a whole number";
      const n = Number(value);
      if (def.min != null && n < def.min) return `must be ≥ ${def.min}`;
      if (def.max != null && n > def.max) return `must be ≤ ${def.max}`;
      return null;
    }
    default: {
      const re = FORMAT_PATTERNS[def.type];
      if (re && !re.test(value)) return `expected ${formatHint(def.type)}`;
      return null;
    }
  }
}

export function formatHint(type: OmsFieldDef["type"]): string {
  switch (type) {
    case "date":
      return "M/D/YYYY";
    case "time":
      return "h[:mm]AM|PM, e.g. 12PM";
    case "percent":
      return "0.000%";
    case "mm":
      return "millions with 3 decimals, e.g. 500.000";
    case "price":
      return "3 decimals, e.g. 99.500";
    case "integer":
      return "a whole number";
    case "boolean":
      return "Yes or No";
    default:
      return "text";
  }
}

/** All problems across a record, keyed by field key. Empty object when clean. */
export function validateFields(fields: FieldValues): Record<string, string> {
  const problems: Record<string, string> = {};
  for (const f of OMS_FIELDS) {
    const p = validateFieldValue(f, fields[f.key] ?? "");
    if (p) problems[f.key] = p;
  }
  return problems;
}

function csvCell(v: string): string {
  const s = v ?? "";
  return /[",\r\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

/** Header row of labels + one data row, "\n" line endings, RFC 4180 quoting. */
export function toCsv(fields: FieldValues): string {
  const header = OMS_FIELDS.map((f) => csvCell(f.label)).join(",");
  const row = OMS_FIELDS.map((f) => csvCell(fields[f.key] ?? "")).join(",");
  return `${header}\n${row}\n`;
}

/** Keys whose current value differs from the original (for the review diff view). */
export function changedKeys(original: FieldValues, current: FieldValues): string[] {
  return OMS_FIELDS.map((f) => f.key).filter(
    (k) => (original[k] ?? "") !== (current[k] ?? ""),
  );
}
